apps/src/lib/node/ledger/shell/block_alloc/states.rs

```diff
@@ -64,36 +64,41 @@ pub struct BuildingEncryptedTxBatch<Mode> {
 /// Allow block proposals to include encrypted txs.
 ///
 /// For more info, read the module docs of
-/// [`crate::node::ledger::shell::block_space_alloc::states`].
+/// [`crate::node::ledger::shell::block_alloc::states`].
 pub enum WithEncryptedTxs {}
 
 /// Prohibit block proposals from including encrypted txs.
 ///
 /// For more info, read the module docs of
-/// [`crate::node::ledger::shell::block_space_alloc::states`].
+/// [`crate::node::ledger::shell::block_alloc::states`].
 pub enum WithoutEncryptedTxs {}
 
-/// Try to allocate a new transaction on a [`BlockSpaceAllocator`] state.
+/// Try to allocate a new transaction on a [`BlockAllocator`] state.
 ///
 /// For more info, read the module docs of
-/// [`crate::node::ledger::shell::block_space_alloc::states`].
+/// [`crate::node::ledger::shell::block_alloc::states`].
 pub trait TryAlloc {
-    /// Try to allocate space for a new transaction.
-    fn try_alloc(&mut self, tx: &[u8]) -> Result<(), AllocFailure>;
+    type Resource<'tx>;
+
+    /// Try to allocate resources for a new transaction.
+    fn try_alloc(
+        &mut self,
+        resource_required: Self::Resource<'_>,
+    ) -> Result<(), AllocFailure>;
```

Comment:
Ah, this can't actually be done because some times we need the `Resource` to be of type `BlockResources` (so both gas and len) which cannot implement the proposed `Resource` trait given that a single `TxBin` can track only a single resource

Reply:
you can add a `Resource` impl for `BlockResources` I guess

Reply:
I could but the fact is that `TxBin` is now generic over a `R: Resource`, but the bin itself cannot actually track more than one resource. It's not a problem if we never instantiate a `TxBin<BlockResource>` but I don't know, it still lefts this possibility open